packages/core/src/sync/index.ts

```diff
@@ -936,243 +817,442 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
+  params.common.logger.debug({
+    service: "sync",
+    msg: `Initialized '${params.network.name}' extract query for timestamps [${decodeCheckpoint(params.from).blockTimestamp}, ${decodeCheckpoint(params.to).blockTimestamp}]`,
+  });
+
+  for await (const syncCheckpoint of bufferAsyncGenerator(
+    params.localSyncGenerator,
+    Number.POSITIVE_INFINITY,
+  )) {
+    let consecutiveErrors = 0;
+    while (cursor < min(syncCheckpoint, params.to)) {
+      const estimateCheckpoint = encodeCheckpoint({
+        ...ZERO_CHECKPOINT,
+        chainId: BigInt(params.network.chainId),
+        blockTimestamp: Math.min(
+          decodeCheckpoint(cursor).blockTimestamp + estimateSeconds,
+          MAX_CHECKPOINT.blockTimestamp,
         ),
-      ),
-    );
+      });
+      const to = min(syncCheckpoint, estimateCheckpoint, params.to);
+      try {
+        const { events, cursor: queryCursor } =
+          await params.syncStore.getEvents({
+            filters: params.sources.map(({ filter }) => filter),
+            from: cursor,
+            to,
+            limit: params.limit,
+          });
 
-    if (completedIntervals.length === 0) return undefined;
+        params.common.logger.debug({
+          service: "sync",
+          msg: `Extracted ${events.length} '${params.network.name}' events for timestamps [${decodeCheckpoint(cursor).blockTimestamp}, ${decodeCheckpoint(queryCursor).blockTimestamp}]`,
+        });
 
-    const earliestCompletedInterval = completedIntervals[0]!;
-    if (earliestCompletedInterval[0] !== (filter.fromBlock ?? 0)) {
-      return undefined;
-    }
-    return earliestCompletedInterval[1];
-  });
+        estimateSeconds = estimate({
+          from: decodeCheckpoint(cursor).blockTimestamp,
+          to: decodeCheckpoint(queryCursor).blockTimestamp,
+          target: params.limit,
+          result: events.length,
+          min: 10,
+          max: 86_400,
+          prev: estimateSeconds,
+          maxIncrease: 1.08,
+        });
 
-  const minCompletedBlock = Math.min(
-    ...(latestCompletedBlocks.filter(
-      (block) => block !== undefined,
-    ) as number[]),
-  );
+        params.common.logger.debug({
+          service: "sync",
+          msg: `Updated '${params.network.name}' extract query estimate to ${estimateSeconds} seconds`,
+        });
 
-  /**  Filter i has known progress if a completed interval is found or if
-   * `_latestCompletedBlocks[i]` is undefined but `sources[i].filter.fromBlock`
-   * is > `_minCompletedBlock`.
-   */
-  if (
-    latestCompletedBlocks.every(
-      (block, i) =>
-        block !== undefined ||
-        (sources[i]!.filter.fromBlock ?? 0) > minCompletedBlock,
-    )
-  ) {
-    return _eth_getBlockByNumber(requestQueue, {
-      blockNumber: minCompletedBlock,
-    });
-  }
+        consecutiveErrors = 0;
+        cursor = queryCursor;
+        yield { events, checkpoint: cursor };
+      } catch (error) {
```

Comment:
I think we should almost certainly log this error, right? Or is it getting logged elsewhere already?